### ISSUE TITLE
"Advise" in the ipc sample will stop working after a few iterations when using TCP for IPC -- wxMSW

### ISSUE THREAD

**Issue by jpmattia** [wxWidgets/wxWidgets#24639]
### Description
On windows: "Advise" in the ipc sample will stop working after a few iterations when using TCP for IPC
ie #define wxUSE_DDE_FOR_IPC 0

#### Observed behavior and how to reproduce

1. Start both ipcserver and ipcclient.
2. Start the server on the server gui, click connect on the client gui.
3. Click Start Advise on the client gui
4. Click Advise on the server gui a few times

After one or more clicks of "Advise" on the server, the message stops being passed back to the client. 

5. click Stop Advise and disconnect on the client, and usually I see a crash. When running ipcclient under gdb, here is the backtrace:

```
(gdb) bt
#0  0x00007ff7c5a10470 in wxSocketBase::LastError() const ()
#1  0x00007ff7c5a0eac0 in wxSocketInputStream::OnSysRead(void*, unsigned long long) ()
#2  0x00007ff7c5a8f34f in wxInputStream::Read(void*, unsigned long long) ()
#3  0x00007ff7c5a0da29 in wxTCPEventHandler::Client_OnRequest(wxSocketEvent&) ()
#4  0x00007ff7c5afbfa4 in wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&) ()
#5  0x00007ff7c5afc13b in wxEventHashTable::HandleEvent(wxEvent&, wxEvtHandler*) ()
#6  0x00007ff7c5afc84b in wxEvtHandler::TryHereOnly(wxEvent&) ()
#7  0x00007ff7c5afc88c in wxEvtHandler::ProcessEventLocally(wxEvent&) ()
#8  0x00007ff7c5afc992 in wxEvtHandler::ProcessEvent(wxEvent&) ()
#9  0x00007ff7c5afd77d in wxEvtHandler::ProcessPendingEvents() ()
#10 0x00007ff7c5a1a4df in wxAppConsoleBase::ProcessPendingEvents() ()
#11 0x00007ff7c5a459d7 in wxEventLoopManual::ProcessEvents() ()
#12 0x00007ff7c5a45a9c in wxEventLoopManual::DoRun() ()
#13 0x00007ff7c5a457bb in wxEventLoopBase::Run() ()
#14 0x00007ff7c5a1bfbc in wxAppConsoleBase::MainLoop() ()
#15 0x00007ff7c5a6d460 in wxEntryReal(int&, wchar_t**) ()
#16 0x00007ff7c57c12ee in __tmainCRTStartup () at C:/M/B/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:267
#17 0x00007ff7c57c13e6 in WinMainCRTStartup () at C:/M/B/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:157
```

### Platform and version information

- wxWidgets version  you use: 3.1.3
- wxWidgets port  you use: wxMSW
- OS  and its version: Windows 10, MinGW, gcc.exe (Rev3, Built by MSYS2 project) 13.2.0

I realize MinGW under windows is a little different compared to the average/median developer, but I'm a little baffled as to why that would make a difference to the sample code. This sample is working for everyone else?

### Kudos and Thanks

I know this is a bug report, but hopefully I'm allowed one line to express my thanks to all of the wxWidgets contributors!

**Comment by vadz** [comment]
Sorry, why do you think this is MinGW-specific, did you test the sample built with MSVC and does it work there?

Also, does this only happen when using TCP (so a `wxSocket` bug, probably) or also when using DDE (more likely to be a bug in `wxConnection` itself)?

I'm afraid I'm not going to have enough time to look into this myself in the observable future, but perhaps knowing the answers to the questions above might give some ideas...

P.S. Thanks for your kind words!

**Comment by jpmattia** [comment]
My MinGW comment was more intuition-based: I'd be a little surprised that such an obvious bug in the ipc sample would not have been noticed by now with the general public, so I was taking a wild guess that it must be some combination of wxMSW + TCP + MinGW, and MinGW is probably the least used item in that set. But you're right, I have no good reason to suspect it.

I unfortunately don't have a ready setup for MSVC: I tried opening VS2022 just now, but the build doesn't find wx/setup.h. I don't use MSVC (I'm not terrible knowledgeable about microsoft compilers, and I'm only on the msw platform because my customers are there) so at the moment, I don't know how to configure it to find the correct includes for compile and libs for link. (Normal IDE is eclipse on this machine.)

I did take a moment to recompile the sample using DDE: It works fine as near as I can tell. So maybe a problem with wxSocket. It does suggest that fastest path to resolution is not use TCP, but I'm worried about support for DDE (a quick googling of [Is DDE still supported?](https://www.google.com/search?sca_esv=029f1710fb0d870d&sca_upv=1&q=Is+DDE+still+supported%3F) gives me heartburn about sending the end product to customers.)

Another thing I noticed: The following sequence works:

1. open ipcclient and ipcserver
2. start server
3. connect 
4. start advise
5. advise
6. stop advise
7. disconnect

Steps 3-7 can be performed repeatedly without losing the advise message, and without crashing on disconnect. Or at least, it works for longer and I haven't hit it enough times.

One last observation: Repeated "Request" operations work pretty well, but eventually an assert fires in common/socket.cpp (line 210 in v3.1.3):         wxASSERT_MSG( !m_socket->m_reading, "read reentrancy?" );  There is an old issue #11528, but I noticed that someone in 2019 made the comment about the sample still showing the bug:
https://github.com/wxWidgets/wxWidgets/issues/11528#issuecomment-1010749352  I don't know if the year is right (did the trac migration post-date comments?) but I'm worried I've stumbled upon an old unsolved issue.

Anyway, I'm willing to try and look at wxSocket but I don't know I'll be of much use because 1. I didn't know of its existence before seeing it on the backtrace above, and 2. my knowledge of the TCP/IP stack is terrible and dangerous to others.  But if any of the above observations suggest something to you, let me know.

Again, many thanks.

**Comment by vadz** [comment]
I think few people use wxIPC at all and very few use TCP under Windows (DDE remains the de facto standard for the kind of things wxIPC is usually used for, e.g. remotely controlling another application instance), so I am not surprised at all that it's broken, unfortunately.

And yes, it does look like the other issue, I must have tested the fix in 2010 but maybe something got broken again later. The comment clearly made a typo in the other ticket number, so I have no idea whether it's exactly the same issue or not, but it could well be...

**Comment by jpmattia** [comment]
> _(DDE remains the de facto standard for the kind of things wxIPC is usually used for,...)_

Fair enough, I'll switch to DDE out of expediency. Let me know if I should close this issue (or you can), or set its disposition differently if it's somehow useful to getting wxSockets looked at.

**Comment by jpmattia** [comment]
OK, I've been poking around wxSocket with the IPC sample using TCP.  Under the latest release 3.2.5 on windows 10, I added 

`wxLogMessage("wxSocketBase::Read begin");`

to the first line of wxSocketBase::Read. Adding that line seems to solve the hang issue.  I do not yet understand how that might influence winsock or the eventloop that wxWidgets uses for the sockets, but does it suggest anything to you (or any of the wxWidget gurus) about why that would eliminate the IPC hanging behavior?

**Comment by vadz** [comment]
Logging a message wakes up the main thread, so this might indicate that we need a call to `wxWakeUpIdle()` somewhere, but OTOH moving the mouse would wake up the main thread too, so it's probably not that simple.

**Comment by jpmattia** [comment]
OK, I've just spent a bunch of time over the last week, doing a deep dive on the TCP version of the IPC implementation.

I believe the low level source of the problem is the behavior from unexpected windows sockets errors: There are errors
wxSocketImplMSW::GetLastError classifies as wxSOCKET_IOERR that would be solved by DoWaitWithTimeout() (as with
wxSOCKET_WOULDBLOCK in wxSocketBase::DoRead), and moreover the wxSOCKET_IOERR errors are essentially ignored by the
wxIPCSocketStreams that wxTCPConnection uses for socket IO.

To explain the issue in concrete terms, I'm going to write about the client side, and I note this example is for latest
v3.2.5 code. For the case of the client gui, you can see the problem by:

1. Add wxLogMessage to wxSocketImplMSW::GetLastError (src/msw/sockmsw.cpp) for when the default case is triggered.
2. Print out the error number that was found from WSAGetLastError()
3. By clicking "Start Advise" or "Execute", You will sometimes (often?) find ERROR_ACCESS_DENIED (error 5) and
   WSAECONNABORTED (Error 10053, An established connection was aborted by the software in your host machine).

I have no idea why errors 5 or 10053 are being triggered; googling these errors point to firewalls/antivirus etc.  But
they cause the following cascade to occur:

1. wxSOCKET_IOERR is returned to the m_impl->Read call inside of the wxSocketBase::DoRead
2. SetError(wxSOCKET_IOERR) is then set, and the wxSocketBase::DoRead method returns 0 as the number of bytes read.
3. Then in wxSocketBase::Read, m_lcount_read and m_lcount are set to 0 bytes read.
4. wxSocketInputStream::OnSysRead returns 0 bytes read, and sets its m_lasterror to wxSTREAM_READ_ERROR.
5. wxInputStream::Read looks at the return from OnSysRead, and it returns because 0 bytes means no more data is
   available.  There was no error-check on the wxSocketInputStream.
6. wxDataInputStream::Read8() then returns the wxUint8 value of buf, but note buf has not had any data read into it. In
   my debug case, the uninitialized value is 0.
7. wxIPCSocketStreams::Read8 uses that uninitialized value of 0 and returns it. The IPCstreams never check for a read
   error that I see.
8. wxTCPConnection::StartAdvise receives that zero, and returns false because 0 is not equal to IPC_ADVISE_START
9. The ipc sample code never checks the result of StartAdvise, so we miss that StartAdvise returned false. The server side activated its "Advise" button, so it looks like everything went fine.
      
I think that later reads on the socket will find the unread byte (=IPC_ADVISE_START). This leads to a situation where
the expected data reads are misaligned to the actual data reads. Often you will see "Unexpected IPC_FAIL" or other
unexpected bytes that come out on the client. This dump has the effect of aligning subsequent reads for many gui actions 
after the dump, so it looks like the IPC client works for a while. I don't have a solid grasp of why the
wxSocketReadGuard gets triggered, but I think the misalignment means one read is stuck in DoWaitWithTimeout() inside
wxSocketBase::DoRead, when a subsequent read is initiated.

So now the question is how to best fix the problem. I think there is a smaller fix needed to make the immediate code
work to get my project out the door, and probably a bigger fix to make IPC robust and future proof.

Smaller fix: wxSocketImplMSW::GetLastError should add cases for ERROR_ACCESS_DENIED (5) and error WSAECONNABORTED
(10053) at the least. If those cases return wxSOCKET_WOULDBLOCK, DoRead waits properly for the data after the initial
errors. (I've mostly verified this fix makes the ipc sample work.)

It seems unclean to return wxSOCKET_WOULDBLOCK for WSAECONNABORTED or ERROR_ACCESS_DENIED, so perhaps the right thing is to create a new socket error like wxSOCKET_IOERR_TRYAGAIN and change wxSocketBase::DoRead accordingly so that DoWaitWithTimeout is called, just like wxSOCKET_WOULDBLOCK would do. More generally, are there cases for wxMSW where wxSOCKET_IOERR should not just retry after waiting?

The bigger fix I believe is needed: A return value of wxSOCKET_IOERR is mostly ignored by wxTCPConnection. For the case
I outlined above, a 0 was returned by Read8(), but there was no indication that a problem occurred. Given the 5 and
10053 errors, I have to wonder if there are other errors that are possible, or might happen in the future.  My guess is
that checking whether an error has been set by the socket needs to be incorporated more thoroughly in order to make it
robust. Perhaps disconnect should be called after wxSOCKET_IOERR, but maybe that's drastic.

Also in the category of drastic: Given the way wxTCPConnection operates, I'm not sure I see the value of wxIPCstreams,
and I'm wondering whether wxTCPConnection should eliminate them in favor of using wxSocket directly.

I can take a crack at the smaller fix, and I think it's quick. I don't know what side effects my proposed fix would
cause, so any guidance on that front is welcome.  The bigger fix is another story, and I'd definitely need some guidance
as to how to implement that. It would certainly have to wait until I fill some other obligations.

In any case Vadim, many thanks for your help and suggestions through this debug process.
JP

PS. Other nits that I found in the code that someone more knowledgable could chime in on:

1. There are two calls to WSAGetLastError (via wxSocketImplMSW::GetLastError)
   - First in wxSocketImpl::Read after RecvStream is called
   - Second in wxSocketBase::DoRead when trying to figure out if the error is wxSOCKET_WOULDBLOCK

   I think the right thing is to store the first call, in case any intervening error occurs.

2. In sckipc.cpp, wxIPCSocketStreams dates from the time MTU was 1500 bytes max, so m_bufferedOut should probably be
   updated to 64K - overhead (= 65483? Someone with modern TCP/IP knowledge should check me.)

3. The wxSocket timeout is set to the default of 10 minutes. For IPC, I'd vote for 60 seconds or fewer, in order that
   IPC looks responsive for the case of a timeout. Ideally, it should be settable from wxConnection.

4. In the IPC client.cpp sample the sizer for the wxTextControl should expand vertically as well as horizontally, imho.

**Comment by jpmattia** [comment]
I think I've identified a couple more bugs, and since these are in src/common/sckipc.cpp, I doubt they're specific to wxMSW builds.

First: As near as I can tell, there is a race condition when the server is sending Advise data and the client has a Request.  Imagine the server is sending Advise data regularly (every 0.1s) and the client does a request. In wxTCPConnection::Request, the first byte must be IPC_REQUEST_REPLY.  If the server has snuck in an Advise packet, then the first byte will be IPC_ADVISE, so wxTCPConnection::Request will abort reading (leaving the Advise data to be read next, causing data misalignment.

Second: If there is a data misalignment as in the previous comment, there is a probability that the read in wxTCPEventHandler::Client_OnRequest (const int msg = streams->Read8();) will find a byte corresponding to IPC_FAIL. This byte will set error=true, so that IPC_FAIL is sent back.  It appears that the other side will sometimes be in Client_OnRequest, receive the IPC_FAIL and then send another IPC_FAIL back to the initial side...  ad infinitum, or until some other action interrupts the loop.

As for fixes- For the first issue: I think Request+Advise is a unique case, in that there are no other combinations where the arrival of Advise data will cause an issue. As such, wxTCPConnection::Request should 
  - explicitly check for IPC_ADVISE
  - call a method to dispatch the read/OnAdvise if it is present
  - try again to read for IPC_REQUEST_REPLY

For the first issue, not sending back IPC_FAIL is probably right (don't set error=true for that switch case). If sending a reply is somehow very necessary, then probably creating another enum IPC_FAIL_ACK and sending that (and not replying to it) is probably the right thing.

Since misalignment of the data keeps arising, I'm also tempted to suggest that the IPC_XXX enum should use fairly random 32 bit values, which will reduce the probability that data misalignment causes erroneous calls within wxTCPEventHandler::Client_OnRequest or other methods. Many of the Read8/Write8 methods would change to Read32/Write32.  

Any comments/suggestions welcome.

**Comment by jpmattia** [comment]
I think I've identified another bug, and since this is in src/common/sckipc.cpp, I doubt it's specific to wxMSW builds.

As near as I can tell, there is a race condition when the server is sending Advise data and the client has a Request.  Imagine the server is sending Advise data regularly (every 0.1s) and the client does a request. In wxTCPConnection::Request, the first byte must be IPC_REQUEST_REPLY.  If the server has snuck in an Advise packet, then the first byte will be IPC_ADVISE, so wxTCPConnection::Request will abort reading (leaving the Advise data to be read next, causing data misalignment.)

As for a fix: I think Request+Advise is a unique case, in that there are no other combinations where the arrival of Advise data will cause an issue. As such, wxTCPConnection::Request should 
  - explicitly check for IPC_ADVISE
  - call a method to dispatch the read/OnAdvise if it is present
  - try again to read for IPC_REQUEST_REPLY

Since misalignment of the data keeps arising, I'm also tempted to suggest that the IPC_XXX enum should use fairly random 32 bit values, which will reduce the probability that data misalignment causes erroneous calls within wxTCPEventHandler::Client_OnRequest or other methods. Many of the Read8/Write8 methods would have to change to Read32/Write32.  

Comments/suggestions welcome.

**Comment by vadz** [comment]
Thanks a lot for spending time to debug this and sorry for the delay with reply!

I had no idea that we were getting `ACCESS_DENIED` and `WSAECONNABORTED`, but it seems clear that we must handle them. But handling them as `wxSOCKET_WOULDBLOCK` doesn't seem right to me, at least for the former, they really look like fatal errors to me. Maybe this will still result in some errors in the sample, but at least they will be visible and won't prevent future operations from working.

I.e. IMO the real problem is lack of error handling in `wxTCPConnection` and if you could please add it, this would be great. If you can't do this, or at least not right away, could we at least detect the error and close the connection? Hopefully this should be relatively simple to do...

Concerning removing `wxIPCSocketStreams`: AFAICS they are used to allow reusing `wxDataInputStream::Read32()` and `wxDataOutputStream::Write32()`, otherwise I indeed don't see any particular value in them. If removing them would make things simpler, I definitely wouldn't regret them.

As for the other questions at the end of the last before one comment:

1. I agree, thanks for noticing this. AFAICS we just need to replace `m_impl->GetLastError()` with `m_impl->GetError()` in both `DoRead()` and `DoWrite()`. Please don't hesitate to make a PR fixing this if you can test it and confirm that this doesn't have any unexpected effects.
1. No objections to increasing the buffer size, but I'm not sure what actually happens when the buffer overflows, does it really result in a network request or will it still be buffered at the socket level? In any case, this doesn't seem that important to me because I'm sure the performance here, with all the intermediary buffers involved, can't be great anyhow.
1. We definitely should reduce the timeout, I think 10 seconds would be more than enough.
1. Sure, again, please don't hesitate submitting a PR changing this.

Finally, in reply to the last comment: this looks like a bad bug, and it's not present just here... AFAICS `IPC_ADVISE` may be read at any time at all, so we should check for it after _every_ call to `Read8()`, shouldn't we?

**Comment by jpmattia** [comment]
I had also come to the conclusion that the race condition is more prevalent than my initial comment, and unless I'm mistaken, I think it's even worse than checking after `Read8()` calls. As near as I can tell, there's no requirement that `Advise()` is running in the same thread as `OnRequest()`.  (I believe that is how it is running in our app.)  So in a race condition between `Advise()` and `Request()`,  there's no guarantee that the bytes immediately following `Read8()` correspond to the same call. It is possible that the packets are interleaved, eg ordered as `IPC_REQUEST_REPLY` : `IPC_ADVISE`: (request-reply bytes) : (advise-info bytes).

As I dug in further last week, I think there is another big issue: There seems to be a built-in assumption that each send-recv  loads packets in a 1:1 way with the sckipc.cpp calls. This is my first experience with the socket level, but it is my understanding that the send-recv stream has no guarantees about how bytes are loaded into the packets. As a consequence, one packet may contain the result of multiple calls. An example of what I mean: 
1. The server sends `IPC_REQUEST_REPLY` + its data, followed by `IPC_ADVISE` followed by its data, and all of it might be stuffed in one packet.
2. Windows sends an async event to the client that the socket has data to be read.
3. `wxSocketBase::NotifyOnStateChange()` puts an event in the client queue to process the data.
4. When that event is handled, Client_OnRequest is called
5. The `IPC_REQUEST_REPLY` msg is read and its data dispatched.
6. Client_OnRequest returns and the `IPC_ADVISE`+data is ignored until some other event triggers.

From the server side, multiple requests might be stuffed into one incoming packet, and the server will process only one for each socket event. Data will accumulate in the socket buffer whenever the multiple calls are stuffed into one packet.

I think this explains some behavior I have been seeing, but I haven't written code to test it out explicitly to be sure. 

The fix I will propose for the above problems
- The request type (`IPC_XXX`), the size, the data format, and the data ptr should be put into a struct. Name this `wxIPCPacket`. The struct is sent to the socket, and the entire struct is written out in one `send` call.
- `Client_OnRequest `and `Server_OnRequest` call `wxIPCPacket::Read()`, which reads data into the struct, and dispatches as it does now based on the request type. They iterate read-dispatch until there is no more data to be read.

I had thought I was going to implement a quick fix so that our app could get released, but I'm now of the opinion that I need to tackle serious changes in sckipc.cpp before release, because the number of possibilities for things to go wrong for customers is not in our comfort zone.

Let me know if you have comments about the above, and I'll be starting on changes later this week.

**Comment by jpmattia** [comment]
[Second try at a reply with corrections/additions.]

I also came to the conclusion that the race condition is more prevalent than my initial comment, and unless I'm mistaken, I think it's even worse than checking after `Read8()` calls. As near as I can tell, there's no requirement that `Advise()` is running in the same thread as `OnRequest()`.  (I believe that is how it is running in our app.)  So in a race condition between `Advise()` and `OnRequest()`,  there's no guarantee that the bytes immediately following `Read8()` correspond to the same call. It is possible that the calls are interleaved, eg the bytes are ordered as `IPC_REQUEST_REPLY` : `IPC_ADVISE`: (request-reply bytes) : (advise-info bytes).

As I dug in further last week, I think there is another big issue: There seems to be a built-in assumption that each send-recv  interaction loads packets in a 1:1 way with the sckipc.cpp calls. This is my first experience with the socket level, but it is my understanding that the send-recv stream has no guarantees about how bytes are sent. As a consequence, one packet may contain the result of multiple calls. An example of what I mean: 
1. The server sends `IPC_REQUEST_REPLY` + its data, followed by `IPC_ADVISE` followed by its data, and all of it might be stuffed in one packet.
2. Windows sends an async event to the client that the socket has data to be read.
3. `wxSocketBase::NotifyOnStateChange()` puts an event in the client queue to process the data.
4. When that event is handled, the `IPC_REQUEST_REPLY` msg is read and its data processed.
6. The `IPC_ADVISE`+data is ignored until some other event triggers.

From the server side, multiple requests might be stuffed into one incoming packet, and the server will process only one for each socket event. Data will accumulate in the socket buffer whenever the multiple calls are stuffed into one packet.

I think this explains some behavior I have been seeing, but I haven't written code to test it out explicitly to be sure. 

The fix I will propose for the above problems
- The request type (`IPC_XXX`), the size, the data format, and the data ptr should be bundled into a struct. Name this `wxIPCPacket`. The struct is sent to the socket, and the entire struct is written out in one `send` call.
- `Client_OnRequest `and `Server_OnRequest` call `wxIPCPacket::Read()`, which reads data into the struct, and dispatches as it does now based on the request type. `OnRequest()` iterates read-dispatch until there is no more data to be read.
- `wxTCPConnection::Request` also needs to make sure there is no more data from recv after it processes IPC_REQUEST_REPLY.

I had thought I was going to implement a quick fix so that our app could get released, but I'm now of the opinion that I need to tackle serious changes in sckipc.cpp before release, because the number of possibilities for things to go wrong for customers is not in our comfort zone.

Let me know if you have comments about the above, and I'll be starting on changes later this week.

**Comment by vadz** [comment]
I think we can assume that a single thread is used because nothing UI-related is thread-safe anyhow and we can pretend, squinting just a little, that all this IPC stuff is "UI-related".

However I don't think there is any problem related to how the OS breaks stuff we send into TCP packets, this is beneath the layer of abstraction we're operating on. At the socket level there is just a stream and it doesn't matter if the `N` bytes we're trying to read from it came in one packet or 10 of them. So things are probably not **that** bad.

**Comment by jpmattia** [comment]
OK, these are my first pull requests on github, so please bear with me as I learn protocol.

Looking back over my previous comment, I’ve used some confusing language (in particular, I overloaded the word “packet”.) Let me try again:

Let’s name the collection of IPC bytes associated with one operation a “IPC message”, to differentiate it from a "TCP/IP packet”.  So one “IPC message” might be IPC_ADVISE _plus all the bytes associated with that advise operation_.  The new bug I found is then: The OS might stuff multiple IPC_ADVISE messages into a single packet. Then, the OS will send a *single* notification to the queue in sckipc.cpp via `wxSocketImpl->NotifyOnStatusChange()` which generates an event, and that when that event is processed, only one of the IPC messages is read before returning.  The remaining IPC messages stay in the OS socket buffer until another notification arrives, and then sckipc.cpp again processes only one message (really from the previous TCP/IP packet), leaving any additional messages still piling up in the OS socket buffer.

So the way we do need to worry about the TCP/IP abstraction is in how it notifies the socket about waiting data. Is there one `wxSocketImpl->NotifyOnStatusChange()` per `send`? Then it's not a problem. On wxMSW, it looks to me like there is a problem, so multiple IPC messages (with all of their associated `send` calls) are likely stuffed into one TCP/IP packet, sckipc.cpp gets only one notification of the TCP/IP packet arriving, and then processes only one of the multiple IPC messages.

In any case, I'll rework sckipc.cpp to remove the streams, and I think it's straightforward to clean up both the race conditions as well as multiple-IPC-messages-per-TCP/IP-notification issues while doing it. (And I hope this doesn't get filed under famous last words.)

**Comment by jpmattia** [comment]
OK my PR for a fix is done, and now I'm looking at auto-tests. The current test (in net/ipc.cpp) is omitted (via preprocessor `#if 0`) with a note about the test hanging in `Connect()` "for unknown reason and this prevents buildbot builds from working so disabling it, but the real problem needs to be fixed, of course."

As near as I can tell, the skeleton there _should_ hang on linux and windows (I don't know OSX) because both the server and the client are running in the same process: The `wxSocketImpl` callbacks for the server socket and the client socket cannot both run simultaneously in the same process. Whatever last callback installed will have priority, and therefore only the server or the client (but not both) will receive the notification for reading the socket.

My plan to fix this is to start a wxServer in a separate process using wxExecute, and then have wxClient make calls out to it. So far, I've verified that if I eliminate IPCTestServer, and then run baseipcserver from the ipc sample, the `Connect` test will run.

I don't know if that approach is suitable for the automated test platform on github, so please give me a heads-up if you see a problem.

**Comment by jpmattia** [comment]
Just added an IPC test to the pull request.  All the tests work on wxMSW. All but one test under wxGTC; the most stringent test fails.  Details in the PR https://github.com/wxWidgets/wxWidgets/pull/24858
